### ISSUE TITLE
Running SELECT COUNT(*) on a large dataset causes loss of interactivity

### ISSUE THREAD

**Issue by rjnn** [MaterializeInc/materialize#7162]
```
materialize=> select count(*) from crew;
  count   
----------
 45458370
(1 row)

Time: 55129.690 ms (00:55.130)
```

While this is running, additional terminals cannot run any queries, nor does CTRL+C return to the terminal - Materialize does not respond until this query finishes, which is a frustrating user experience.

**Comment by frankmcsherry** [comment]
Part of this is surely related to https://github.com/MaterializeInc/materialize/issues/7031 which is the issue that decanting records from an existing arrangement is not fueled. That does explain why it is not easily cancelled (the operator is wedged) but .. I would say maybe not explain why the shell cannot be returned (MZ could in principle know that a result is cancelled and return control, rather than waiting for the dataflow layer to ack the cancellation).

**Comment by frankmcsherry** [comment]
Independently, I'd like to figure out why it takes 55 seconds to count up to 45 million.

**Comment by benesch** [comment]
> (MZ could in principle know that a result is cancelled and return control, rather than waiting for the dataflow layer to ack the cancellation).

Yeah, I've been leery of fixing this because it just papers over the problem. https://github.com/MaterializeInc/materialize/issues/2392#issuecomment-689822683

> if Ctrl+C is hanging, it means trivial queries like `SELECT * FROM view LIMIT 1` are also hanging.

Like, we can give your shell back, but then the _next_ thing you do will get wedged as soon as it hits the dataflow layer.

**Comment by frankmcsherry** [comment]
Your next query might hang, but you could in the meantime get back to updating the catalog, e.g. dropping that view so that on restart it isn't reformed, etc. I think this is both a UX issue and a system issue (we might not want to return control, but also we cannot return control), and we could aim to fix the system issue (among the other system issues to fix that contribute).

**Comment by benesch** [comment]
FWIW the cancellation system was intentionally designed not to early return like that. In Postgres, pressing Ctrl-C only gives you your shell back once the database has actually stopped processing your query. As of today, of course, we don't actually wait for your dataflow to abort before giving you your shell back (see #2392), but the intent was one day we would. 

We could definitely pick different cancellation semantics, but wanted to flag that the historical intent was to make cancellation even _less_ eager, not more. IMO the system issues are:

1. We can't actively cancel dataflows
2. We sometimes lose interactivity

In a hypothetical world where both (1) and (2) were fixed, getting your shell back after pressing Ctrl-C would be pretty responsive and _also_ guarantee that the work you scheduled had stopped.

**Comment by uce** [comment]
After #7171 is merged we should confirm that this issue is fixed.

**Comment by uce** [comment]
@frankmcsherry I tried to confirm that this improved with dcd206e1f (includes #7171), but I actually ran into a (probably different) problem.

On dcd206e1f, when I cancel the query I get stuck with `Cancel request sent` and I don't get control back whereas I do get the expected `ERROR:  canceling statement due to user request` eventually on da38e0c97 (around the time Arjun created the issue).

I don't have time to look into what's happening unfortunately, it might totally be me screwing something up.
